packages/babel-parser/src/plugins/estree.ts

```diff
@@ -121,6 +123,24 @@ export default (superClass: typeof Parser) =>
       return stmt;
     }
 
+    /**
+     * The TS-ESLint always define optional AST properties, here we provide the
+     * default value for such properties immediately after `finishNode` was invoked.
+     * This hook will be implemented by the typescript plugin.
+     *
+     * Note: This hook should be manually invoked when we change the `type` of a given AST
+     * node, to ensure that the optional properties are correctly filled.
+     * @param node The AST node finished by finishNode
+     */
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    fillOptionalPropertiesForTSESLint(node: NodeType) {}
+
+    castNodeAs<T extends N.Node>(node: N.Node, type: T["type"]): T {
```

Comment:
Nit, naming suggestion
```suggestion
    castNodeTo<T extends N.Node>(node: N.Node, type: T["type"]): T {
```